universum/lib/gravity.py

```diff
@@ -39,69 +43,80 @@ def __setattr__(self, key, value):
+    def __get__(self, obj: 'Module', objtype: Any = None) -> 'ModuleSettings':
         return ModuleSettings(self.cls, obj.main_settings)
 
-    def __set__(self, obj, settings):
+    def __set__(self, obj: 'Module', settings: 'Settings'):
         setattr(obj.main_settings, self.cls.__name__, settings)
 
 
-class Module:
-    def __new__(cls, main_settings, *args, **kwargs):
-        instance = super(Module, cls).__new__(cls)
+class Module():
+    settings: ClassVar['Settings']
+    main_settings: 'ModuleSettings'
+    def __new__(cls: Type['Module'], main_settings: 'ModuleSettings', *args, **kwargs) -> 'Module':
+        instance: Module = super(Module, cls).__new__(cls)
         instance.main_settings = main_settings
         return instance
 
 
-def construct_component(klass, main_settings, *args, **kwargs):
+ComponentType = TypeVar('ComponentType', bound='Module')
+
+def construct_component(cls: Type[ComponentType], main_settings: 'ModuleSettings', *args, **kwargs) -> ComponentType:
     if not getattr(main_settings, "active_modules", None):
         main_settings.active_modules = dict()
 
-    if klass not in main_settings.active_modules:
-        klass.settings = Settings(klass)
-        instance = klass.__new__(klass, main_settings=main_settings)
-        instance.__init__(*args, **kwargs)
-        main_settings.active_modules[klass] = instance
-    return main_settings.active_modules[klass]
+    if cls not in main_settings.active_modules:
+        cls.settings = Settings(cls)
+        instance: 'Module' = cls.__new__(cls, main_settings=main_settings)
+        # https://github.com/python/mypy/blob/master/mypy/checkmember.py#180
+        # Accessing __init__ in statically typed code would compromise
+        # type safety unless used via super().
+        instance.__init__(*args, **kwargs)  # type: ignore
+        main_settings.active_modules[cls] = instance
+    return cast(ComponentType, main_settings.active_modules[cls])
+
 
+DependencyType = TypeVar('DependencyType', bound='Module')
 
-class Dependency:
-    def __init__(self, klass):
-        self.klass = klass
+class Dependency(Generic[DependencyType]):
+    def __init__(self, cls: Type[DependencyType]) -> None:
+        self.cls = cls
 
-    def __get__(self, instance, owner):
-        def constructor_function(*args, **kwargs):
-            return construct_component(self.klass, instance.main_settings, *args, **kwargs)
+    def __get__(self, instance: DependencyType, owner: Any) -> Callable[..., DependencyType]:
+        def constructor_function(*args, **kwargs) -> DependencyType:
+            return construct_component(self.cls, instance.main_settings, *args, **kwargs)
         return constructor_function
 
 
-def get_dependencies(klass, result=None, parent=None):
+def get_dependencies(cls: Type[Module], result: Optional[List[Type[Module]]] = None,
+                     parent: Optional[Dict[Type[Module], Type[Module]]] = None) -> List[Type[Module]]:
```

Comment:
You can safely remove `parent` argument instead of annotating it.
We should have removed it during the refactoring, but didn't do it by mistake.